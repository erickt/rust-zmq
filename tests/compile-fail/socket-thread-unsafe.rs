extern crate zmq;

use std::thread;

macro_rules! t {
    ($e:expr) => (
        $e.unwrap_or_else(|e| { panic!("{} failed with {:?}", stringify!($e), e) })
    )
}

fn main() {
    let mut context = zmq::Context::new();
    let socket = t!(context.socket(zmq::REP));
    let s = &socket;
    let t = thread::spawn(move || {  //~ ERROR he trait bound `*mut std::os::raw::c_void: std::marker::Sync` is not satisfied [E0277]
        t!(s.bind("tcp://127.0.0.1:12345"))
    });
    socket.send(b"ABC", 0);
    t.join().unwrap();
}
