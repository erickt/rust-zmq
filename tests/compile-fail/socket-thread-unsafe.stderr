error[E0277]: `*mut c_void` cannot be shared between threads safely
  --> tests/compile-fail/socket-thread-unsafe.rs:13:27
   |
13 |       let t = thread::spawn(move || {
   |  _____________-------------_^
   | |             |
   | |             required by a bound introduced by this call
14 | |         t!(s.bind("tcp://127.0.0.1:12345"))
15 | |     });
   | |_____^ `*mut c_void` cannot be shared between threads safely
   |
   = help: within `Socket`, the trait `Sync` is not implemented for `*mut c_void`
   = note: required because it appears within the type `Socket`
   = note: required for `&Socket` to implement `Send`
note: required because it's used within this closure
  --> tests/compile-fail/socket-thread-unsafe.rs:13:27
   |
13 |     let t = thread::spawn(move || {
   |                           ^^^^^^^
note: required by a bound in `spawn`
  --> $RUST/std/src/thread/mod.rs
